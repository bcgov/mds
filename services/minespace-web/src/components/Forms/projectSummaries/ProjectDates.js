/* eslint-disable */
import React from "react";
import PropTypes from "prop-types";
import { Field } from "redux-form";
import { maxLength } from "@common/utils/Validate";
import { renderConfig } from "@/components/common/config";

const propTypes = {};

export const ProjectDates = (props) => (
  <>
    <h1>Project Dates</h1>
    <Field
      id="expected_draft_irt_submission_date"
      name="expected_draft_irt_submission_date"
      label="When do you anticipate submitting a draft IRT? (optional)"
      placeholder="Please select date"
      component={renderConfig.DATE}
    />
    <Field
      id="expected_permit_application_date"
      name="expected_permit_application_date"
      label="When do you anticipate submitting a permit application? (optional)"
      placeholder="Please select date"
      component={renderConfig.DATE}
    />
    <Field
      id="expected_permit_receipt_date"
      name="expected_permit_receipt_date"
      label="When do you hope to receive your permit/amendment(s)? (optional)"
      placeholder="Please select date"
      component={renderConfig.DATE}
    />
    <Field
      id="expected_project_start_date"
      name="expected_project_start_date"
      label="When do you anticipate starting work on this project? (optional)"
      placeholder="Please select date"
      component={renderConfig.DATE}
    />
  </>
);

ProjectDates.propTypes = propTypes;

export default ProjectDates;
