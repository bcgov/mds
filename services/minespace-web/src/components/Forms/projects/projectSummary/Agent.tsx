import React, { FC, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Field, change, getFormValues } from "redux-form";
import { Col, Form, Row, Typography } from "antd";
import { FORM } from "@mds/common";
import RenderField from "@mds/common/components/forms/RenderField";
import RenderRadioButtons from "@mds/common/components/forms/RenderRadioButtons";
import RenderSelect from "@mds/common/components/forms/RenderSelect";
import {
  email,
  maxLength,
  phoneNumber,
  postalCodeWithCountry,
  required,
  requiredRadioButton,
} from "@mds/common/redux/utils/Validate";
import { getDropdownProvinceOptions } from "@mds/common/redux/selectors/staticContentSelectors";

export const Agent: FC = () => {
  const dispatch = useDispatch();
  const formValues = useSelector(getFormValues(FORM.ADD_EDIT_PROJECT_SUMMARY));
  const { agent = {}, is_agent = false } = formValues;
  const { party_type_code, address = {} } = agent ?? {};
  const { address_type_code } = address ?? {};
  const isInternational = address_type_code === "INT";

  const provinceOptions = useSelector(getDropdownProvinceOptions);
  // currently no endpoints, etc, for address_type_code
  const countryOptions = [
    { value: "CAN", label: "Canada" },
    { value: "USA", label: "United States" },
    { value: "INT", label: "International" },
  ];

  // set a value for party type code because required validation doesn't show
  if (!party_type_code) {
    dispatch(change(FORM.ADD_EDIT_PROJECT_SUMMARY, "agent.party_type_code", "ORG"));
  }

  useEffect(() => {
    if (party_type_code === "ORG") {
      dispatch(change(FORM.ADD_EDIT_PROJECT_SUMMARY, "agent.first_name", null));
    }
  }, [party_type_code]);

  useEffect(() => {
    if (address_type_code === "INT") {
      dispatch(change(FORM.ADD_EDIT_PROJECT_SUMMARY, "agent.address.sub_division_code", null));
    }
  }, [address_type_code]);

  return (
    <Form name={FORM.ADD_EDIT_PROJECT_SUMMARY} layout="vertical">
      <div className="ant-form-vertical">
        <Typography.Title level={3}>Agent</Typography.Title>
        <Typography.Paragraph>
          The Applicant may authorize an Agent to deal with the Ministry directly on future aspects
          of this application. This section must be completed in full if an Agent is used. An Agent
          is a person who is not an employee of the Applicant.
        </Typography.Paragraph>

        <Field
          name="is_agent"
          id="is_agent"
          required
          validate={[requiredRadioButton]}
          label="Are you an agent applying on behalf of the applicant?"
          component={RenderRadioButtons}
        />

        {is_agent && (
          <>
            <Field
              name="agent.party_type_code"
              required
              component={RenderRadioButtons}
              defaultValue={"ORG"}
              customOptions={[
                { label: "Organization", value: "ORG" },
                { label: "Person", value: "PER" },
              ]}
              optionType="button"
            />
            {party_type_code === "ORG" && (
              // TODO: With orgbook integration, this should populate something in orgbook, not agent.party_name
              <Field
                name="agent.party_name"
                label="Agent's Company Legal Name"
                required
                validate={[required, maxLength(100)]}
                component={RenderField}
                help="as registered with the BC Registar of Companies"
              />
            )}
            {party_type_code === "PER" && (
              <Row gutter={16}>
                <Col md={12} sm={24}>
                  <Field
                    name="agent.first_name"
                    label="First Name"
                    component={RenderField}
                    required
                    validate={[required, maxLength(100)]}
                  />
                </Col>
                <Col md={12} sm={24}>
                  <Field
                    name="agent.party_name"
                    label="Last Name"
                    component={RenderField}
                    required
                    validate={[required, maxLength(100)]}
                  />
                </Col>
              </Row>
            )}

            <Row gutter={16}>
              {/* TODO: With Orgbook integration, this should probably be the agent.party_name */}
              {/* <Col md={12} sm={24}>
                <Field
                  name="agent.doing_business_as"
                  label="Doing Business As"
                  component={RenderField}
                  help="if different than the Company Legal Name"
                />
              </Col> */}
              <Col md={12} sm={24}>
                <Field name="agent.job_title" label="Agent's Title" component={RenderField} />
              </Col>
            </Row>

            <Row gutter={16}>
              <Col md={8} sm={19}>
                <Field
                  name="agent.phone_no"
                  label="Contact Number"
                  required
                  validate={isInternational ? [required] : [required, phoneNumber]}
                  component={RenderField}
                />
              </Col>
              <Col md={4} sm={5}>
                <Field name="agent.phone_ext" label="Ext." component={RenderField} />
              </Col>
              <Col md={12} sm={24}>
                <Field
                  name="agent.email"
                  label="Email Address"
                  required
                  validate={[required, email]}
                  component={RenderField}
                />
              </Col>
            </Row>

            <Typography.Title level={4}>Mailing Address</Typography.Title>
            <Row gutter={16}>
              <Col md={19} sm={24}>
                <Field
                  name="agent.address.address_line_1"
                  label="Street"
                  required
                  validate={[required]}
                  component={RenderField}
                />
              </Col>
              <Col md={5} sm={24}>
                <Field name="agent.address.suite_no" label="Unit #" component={RenderField} />
              </Col>
            </Row>

            <Row gutter={16}>
              <Col md={12} sm={24}>
                <Field
                  name="agent.address.address_type_code"
                  label="Country"
                  required
                  validate={[required]}
                  data={countryOptions}
                  component={RenderSelect}
                />
              </Col>

              <Col md={12} sm={24}>
                <Field
                  name="agent.address.sub_division_code"
                  label="Province"
                  required={!isInternational}
                  data={provinceOptions.filter((p) => p.subType === address_type_code)}
                  validate={!isInternational ? [required] : []}
                  component={RenderSelect}
                />
              </Col>
            </Row>

            <Row gutter={16}>
              <Col md={12} sm={24}>
                <Field
                  name="agent.address.city"
                  label="City"
                  required
                  validate={[required]}
                  component={RenderField}
                />
              </Col>
              <Col md={12} sm={24}>
                <Field
                  name="agent.address.post_code"
                  label="Postal Code"
                  component={RenderField}
                  validate={[postalCodeWithCountry(address_type_code)]}
                />
              </Col>
            </Row>
          </>
        )}
      </div>
    </Form>
  );
};
