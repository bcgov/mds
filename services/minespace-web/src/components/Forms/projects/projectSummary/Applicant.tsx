import { Col, Row, Typography } from "antd";
import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Field, getFormValues, change } from "redux-form";
import {
  email,
  maxLength,
  phoneNumber,
  postalCodeWithCountry,
  required,
  requiredRadioButton,
} from "@mds/common/redux/utils/Validate";
import RenderRadioButtons from "@mds/common/components/forms/RenderRadioButtons";
import { CONTACTS_COUNTRY_OPTIONS, FORM, IOrgbookCredential } from "@mds/common";
import OrgBookSearch from "@mds/common/components/parties/OrgBookSearch";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCircleCheck, faCircleX, faSpinner } from "@fortawesome/pro-light-svg-icons";
import { isEmpty } from "lodash";
import { verifyOrgBookCredential } from "@mds/common/redux/actionCreators/orgbookActionCreator";
import RenderField from "@mds/common/components/forms/RenderField";
import { getDropdownProvinceOptions } from "@mds/common/redux/selectors/staticContentSelectors";
import RenderSelect from "@mds/common/components/forms/RenderSelect";
import RenderCheckbox from "@mds/common/components/forms/RenderCheckbox";
import { normalizePhone } from "@common/utils/helpers";

const { Title, Paragraph } = Typography;
interface IVerifiedCredential {
  businessNumber: string;
  registrationStatus: string;
  registriesId: number;
}

const Applicant = () => {
  const dispatch = useDispatch();
  const provinceOptions = useSelector(getDropdownProvinceOptions);
  const [credential, setCredential] = useState<IOrgbookCredential>(null);
  const [verified, setVerified] = useState(false);
  const [checkingStatus, setCheckingStatus] = useState(false);
  const [orgBookOptions, setOrgBookOptions] = useState([]);
  const [verifiedCredential, setVerifiedCredential] = useState<IVerifiedCredential>(null);

  const formValues = useSelector(getFormValues(FORM.ADD_EDIT_PROJECT_SUMMARY));
  const {
    applicant = {},
    is_legal_address_same_as_mailing_address = false,
    is_billing_address_same_as_mailing_address = false,
    is_billing_address_same_as_legal_address = false,
  } = formValues;
  const { party_type_code, address = [], party_orgbook_entity = {} } = applicant ?? {};

  const [mailingAddress, legalAddress, businessAddress] = address || [];
  const isMailingInternational = mailingAddress?.address_type_code === "INT";
  const isLegalInternational = legalAddress?.address_type_code === "INT";
  const isBusinessInternational = businessAddress?.address_type_code === "INT";

  useEffect(() => {
    const defaultPartyTypeCode = party_type_code || "ORG";
    dispatch(
      change(FORM.ADD_EDIT_PROJECT_SUMMARY, "applicant.party_type_code", defaultPartyTypeCode)
    );
  }, []);

  useEffect(() => {
    if (party_orgbook_entity && party_orgbook_entity?.credential_id) {
      const options = [
        { text: party_orgbook_entity.name_text, value: party_orgbook_entity.credential_id },
      ];
      setOrgBookOptions(options);
      const payload = {
        businessNumber: party_orgbook_entity.registration_id || "-",
        registrationStatus: !party_orgbook_entity.registration_status ? "Inactive" : "Active",
        registriesId: party_orgbook_entity.credential_id,
      };
      setVerifiedCredential(payload);
    }
  }, [party_orgbook_entity]);

  useEffect(() => {
    setVerified(false);
    setCheckingStatus(true);
    if (credential) {
      dispatch(verifyOrgBookCredential(credential.id)).then((response) => {
        setVerified(response.success);
        setCheckingStatus(false);
        const payload = {
          businessNumber: credential.topic.source_id || "-",
          registrationStatus: credential.inactive ? "Inactive" : "Active",
          registriesId: credential.id,
        };
        setVerifiedCredential(payload);
      });

      const orgBookEntity = {
        registration_id: credential.topic.source_id,
        registration_status: !credential.inactive,
        registration_date: credential.attributes[0].value,
        name_id: credential.topic.local_name.id,
        name_text: credential.topic.local_name.text,
        credential_id: credential.topic.local_name.credential_id,
        company_alias: null,
      };
      dispatch(
        change(FORM.ADD_EDIT_PROJECT_SUMMARY, "applicant.party_orgbook_entity", orgBookEntity)
      );
    }
    if (party_orgbook_entity && party_orgbook_entity?.credential_id) {
      const options = [
        { text: party_orgbook_entity.name_text, value: party_orgbook_entity.credential_id },
      ];
      setOrgBookOptions(options);
    }
  }, [credential]);

  useEffect(() => {
    setCredential(null);
  }, [party_type_code]);

  const renderStatus = () => {
    let icon = faSpinner;
    let color = undefined;
    let text = undefined;

    if ((!checkingStatus && verified) || verifiedCredential) {
      icon = faCircleCheck;
      color = "#45A766";
      text = "Verified on Orgbook BC";
    } else {
      icon = faCircleX;
      color = "#D8292F";
      text = "Not registered on Orgbook BC";
    }

    return (
      <Paragraph strong className="light">
        <FontAwesomeIcon
          size="lg"
          color={color}
          icon={icon}
          spin={checkingStatus && !verifiedCredential}
          className="margin-medium--right"
        />
        {text}
      </Paragraph>
    );
  };

  const updateAddressSubDivisionCode = (index: number) => {
    const selectedProvince = address?.[index]?.sub_division_code
      ? provinceOptions.find((p) => p.value === address?.[index]?.sub_division_code)
      : {};
    if (
      address?.[index]?.address_type_code === "INT" ||
      selectedProvince?.subType !== address?.[index]?.address_type_code
    ) {
      dispatch(
        change(FORM.ADD_EDIT_PROJECT_SUMMARY, `applicant.address[${index}].sub_division_code`, null)
      );
    }
  };

  useEffect(() => {
    updateAddressSubDivisionCode(0);
  }, [address?.[0]?.address_type_code]);

  useEffect(() => {
    updateAddressSubDivisionCode(1);
  }, [address?.[1]?.address_type_code]);

  useEffect(() => {
    updateAddressSubDivisionCode(2);
  }, [address?.[2]?.address_type_code]);

  return (
    <div className="ant-form-vertical">
      <Title level={3}>Applicant Information</Title>
      <Paragraph>This must be the name of the company or person seeking authorization.</Paragraph>
      <Field
        name="applicant.party_type_code"
        id="applicant.party_type_code"
        required
        validate={[requiredRadioButton]}
        label="Applicant Type"
        component={RenderRadioButtons}
        defaultValue={"ORG"}
        customOptions={[
          { label: "Company", value: "ORG" },
          {
            label: "Individual",
            value: "IND",
          },
        ]}
        optionType="button"
      />
      {party_type_code === "ORG" && (
        <div>
          <Field
            name="applicant.party_name"
            id="applicant.party_name"
            required
            validate={[required]}
            label="Company Legal Name"
            setCredential={setCredential}
            data={orgBookOptions}
            help={"as registered with the BC Registar of Companies"}
            component={OrgBookSearch}
          />
          {verifiedCredential && (
            <div className="table-summary-card">
              {renderStatus()}

              <Paragraph className="light margin-none">
                Business Number: {verifiedCredential.businessNumber}
              </Paragraph>

              <Paragraph className="light margin-none">
                BC Registries ID: {verifiedCredential.registriesId}
              </Paragraph>
              <Paragraph className="light margin-none">
                BC Registration Status {verifiedCredential.registrationStatus}
              </Paragraph>
            </div>
          )}
          <Row gutter={16}>
            <Col md={12} sm={24}>
              <Field
                id="applicant.party_orgbook_entity.company_alias"
                name="applicant.party_orgbook_entity.company_alias"
                label="Doing Business As"
                component={RenderField}
              />
            </Col>

            <Col md={12} sm={24}>
              <Field
                id="applicant.party_orgbook_entity.registration_id"
                name="applicant.party_orgbook_entity.registration_id"
                label="Incorporation Number"
                required
                validate={[required]}
                component={RenderField}
              />
            </Col>
          </Row>
        </div>
      )}
      {party_type_code === "IND" && (
        <Row gutter={16}>
          <Col md={8} sm={24}>
            <Field
              name="applicant.first_name"
              label="First Name"
              required
              validate={[required]}
              component={RenderField}
            />
          </Col>
          <Col md={8} sm={24}>
            <Field name="middle_name" label="Middle Name" component={RenderField} />
          </Col>
          <Col md={8} sm={24}>
            <Field
              name="applicant.party_name"
              label="Last Name"
              required
              validate={[required]}
              component={RenderField}
            />
          </Col>
        </Row>
      )}
      <Row gutter={16}>
        <Col md={8} sm={19}>
          <Field
            name="applicant.phone_no"
            label="Contact Number"
            required
            component={RenderField}
            validate={[phoneNumber, maxLength(12), required]}
            normalize={normalizePhone}
          />
        </Col>
        <Col md={4} sm={5}>
          <Field name="applicant.phone_ext" label="Ext." component={RenderField} />
        </Col>
        <Col md={12} sm={24}>
          <Field
            name="applicant.email"
            label="Email Address"
            required
            validate={[required, email]}
            component={RenderField}
          />
        </Col>
      </Row>
      <Typography.Title level={5}>Mailing Address</Typography.Title>
      <Row gutter={16}>
        <Col md={19} sm={24}>
          <Field
            name="applicant.address[0].address_line_1"
            label="Street"
            required
            validate={[required]}
            component={RenderField}
          />
        </Col>
        <Col md={5} sm={24}>
          <Field name="applicant.address[0].suite_no" label="Unit #" component={RenderField} />
        </Col>
      </Row>
      <Row gutter={16}>
        <Col md={12} sm={24}>
          <Field
            name="applicant.address[0].address_type_code"
            label="Country"
            required
            validate={[required]}
            data={CONTACTS_COUNTRY_OPTIONS}
            component={RenderSelect}
          />
        </Col>
        <Col md={12} sm={24}>
          <Field
            name="applicant.address[0].sub_division_code"
            label="Province"
            required={!isMailingInternational}
            validate={!isMailingInternational ? [required] : []}
            data={provinceOptions.filter(
              (p) => p.subType === applicant?.address?.[0]?.address_type_code
            )}
            component={RenderSelect}
          />
        </Col>
      </Row>

      <Row gutter={16}>
        <Col md={12} sm={24}>
          <Field
            name="applicant.address[0].city"
            label="City"
            required
            validate={[required]}
            component={RenderField}
          />
        </Col>
        <Col md={12} sm={24}>
          <Field
            name="applicant.address[0].post_code"
            label="Postal Code"
            component={RenderField}
            required
            validate={[
              postalCodeWithCountry(applicant?.address?.[0]?.address_type_code),
              maxLength(10),
            ]}
          />
        </Col>
      </Row>
      <Typography.Title level={5}>Legal Address</Typography.Title>
      <Field
        id="is_legal_address_same_as_mailing_address"
        name="is_legal_address_same_as_mailing_address"
        component={RenderCheckbox}
        label="Same as mailing address"
        type="checkbox"
      />
      {!is_legal_address_same_as_mailing_address && (
        <>
          <Row gutter={16}>
            <Col md={19} sm={24}>
              <Field
                name="applicant.address[1].address_line_1"
                label="Street"
                required
                validate={[required]}
                component={RenderField}
              />
            </Col>
            <Col md={5} sm={24}>
              <Field name="applicant.address[1].suite_no" label="Unit #" component={RenderField} />
            </Col>
          </Row>

          <Row gutter={16}>
            <Col md={12} sm={24}>
              <Field
                name="applicant.address[1].address_type_code"
                label="Country"
                required
                validate={[required]}
                data={CONTACTS_COUNTRY_OPTIONS}
                component={RenderSelect}
              />
            </Col>
            <Col md={12} sm={24}>
              <Field
                name="applicant.address[1].sub_division_code"
                label="Province"
                required={!isLegalInternational}
                validate={!isLegalInternational ? [required] : []}
                data={provinceOptions.filter(
                  (p) => p.subType === applicant?.address?.[1]?.address_type_code
                )}
                component={RenderSelect}
              />
            </Col>
          </Row>

          <Row gutter={16}>
            <Col md={12} sm={24}>
              <Field
                name="applicant.address[1].city"
                label="City"
                required
                validate={[required]}
                component={RenderField}
              />
            </Col>
            <Col md={12} sm={24}>
              <Field
                name="applicant.address[1].post_code"
                label="Postal Code"
                component={RenderField}
                required
                validate={[
                  postalCodeWithCountry(applicant?.address?.[1]?.address_type_code),
                  maxLength(10),
                ]}
              />
            </Col>
          </Row>
        </>
      )}
      <Typography.Title level={5}>Billing Address</Typography.Title>
      <Row gutter={16}>
        <Col md={8} sm={24}>
          <Field
            id="is_billing_address_same_as_mailing_address"
            name="is_billing_address_same_as_mailing_address"
            label="Same as mailing address"
            disabled={is_billing_address_same_as_legal_address}
            component={RenderCheckbox}
            type="checkbox"
          />
        </Col>
        <Col md={8} sm={24}>
          <Field
            id="is_billing_address_same_as_legal_address"
            name="is_billing_address_same_as_legal_address"
            label="Same as legal address"
            disabled={is_billing_address_same_as_mailing_address}
            component={RenderCheckbox}
            type="checkbox"
          />
        </Col>
      </Row>
      {!is_billing_address_same_as_mailing_address && !is_billing_address_same_as_legal_address && (
        <>
          <Row gutter={16}>
            <Col md={19} sm={24}>
              <Field
                name="applicant.address[2].address_line_1"
                label="Street"
                required
                validate={[required]}
                component={RenderField}
              />
            </Col>
            <Col md={5} sm={24}>
              <Field name="applicant.address[2].suite_no" label="Unit #" component={RenderField} />
            </Col>
          </Row>
          <Row gutter={16}>
            <Col md={12} sm={24}>
              <Field
                name="applicant.address[2].address_type_code"
                label="Country"
                required
                validate={[required]}
                data={CONTACTS_COUNTRY_OPTIONS}
                component={RenderSelect}
              />
            </Col>
            <Col md={12} sm={24}>
              <Field
                name="applicant.address[2].sub_division_code"
                label="Province"
                required={!isBusinessInternational}
                validate={!isBusinessInternational ? [required] : []}
                data={provinceOptions.filter(
                  (p) => p.subType === applicant?.address?.[2]?.address_type_code
                )}
                component={RenderSelect}
              />
            </Col>
          </Row>

          <Row gutter={16}>
            <Col md={12} sm={24}>
              <Field
                name="applicant.address[2].city"
                label="City"
                required
                validate={[required]}
                component={RenderField}
              />
            </Col>
            <Col md={12} sm={24}>
              <Field
                name="applicant.address[2].post_code"
                label="Postal Code"
                component={RenderField}
                required
                validate={[
                  postalCodeWithCountry(applicant?.address?.[2]?.address_type_code),
                  maxLength(10),
                ]}
              />
            </Col>
          </Row>
        </>
      )}
    </div>
  );
};

export default Applicant;
