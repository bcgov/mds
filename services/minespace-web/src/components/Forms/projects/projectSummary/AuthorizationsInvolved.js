import React, { useState } from "react";
import PropTypes from "prop-types";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { Typography, Checkbox } from "antd";
import { DownOutlined } from "@ant-design/icons";
import { Field, FormSection, change, getFormValues } from "redux-form";
import {
  getTransformedProjectSummaryAuthorizationTypes,
  getDropdownProjectSummaryPermitTypes,
} from "@common/selectors/staticContentSelectors";
import { getFormattedProjectSummary } from "@common/selectors/projectSelectors";
import { renderConfig } from "@/components/common/config";
import * as FORM from "@/constants/forms";
import Callout from "@/components/common/Callout";
import CustomPropTypes from "@/customPropTypes";

const propTypes = {
  formattedProjectSummary: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)])
  ).isRequired,
  change: PropTypes.func.isRequired,
  dropdownProjectSummaryPermitTypes: CustomPropTypes.options.isRequired,
  transformedProjectSummaryAuthorizationTypes: PropTypes.arrayOf(
    PropTypes.objectOf(PropTypes.string)
  ).isRequired,
};

export const AuthorizationsInvolved = (props) => {
  const [checked, setChecked] = useState(
    props.formattedProjectSummary ? props.formattedProjectSummary.authorizationOptions : []
  );
  const handleChange = (e, code) => {
    if (e.target.checked) {
      setChecked((arr) => [code, ...arr]);
    } else {
      setChecked(checked.filter((item) => item !== code));
      props.change(FORM.ADD_EDIT_PROJECT_SUMMARY, code, null);
    }
  };

  const setInitialValues = (authorizationType, formValues) => {
    const currentAuthorizationType = formValues?.authorizations?.find(
      (val) => val?.project_summary_authorization_type === authorizationType
    );
    return currentAuthorizationType?.project_summary_permit_type ?? [];
  };

  const renderNestedFields = (code) => (
    <div className="nested-container">
      {code !== "OTHER" && (
        <>
          <Field
            id="project_summary_permit_type"
            name="project_summary_permit_type"
            fieldName={`${code}.project_summary_permit_type`}
            options={props.dropdownProjectSummaryPermitTypes}
            formName={FORM.ADD_EDIT_PROJECT_SUMMARY}
            formValues={props.formattedProjectSummary}
            change={props.change}
            component={renderConfig.GROUP_CHECK_BOX}
            label="What type of permit is involved in your application?"
            setInitialValues={() => setInitialValues(code, props.formattedProjectSummary)}
          />
        </>
      )}
      <Field
        id={`${code}.existing_permits_authorizations`}
        name="existing_permits_authorizations"
        label={
          <>
            If your application involved a change to an existing permit, please list the numbers of
            the permits involved.
            <br />
            <span className="light--sm">Please separate each permit number with a comma</span>
          </>
        }
        component={renderConfig.FIELD}
      />
    </div>
  );

  return (
    <>
      <Typography.Title level={3}>
        Authorizations potentially involved in the project
      </Typography.Title>
      <Callout message="Please select the authorizations that you anticipate needing for this project, based on your current understanding. This is to assist in planning and may not be the complete list for the final application." />
      {props.transformedProjectSummaryAuthorizationTypes.map((authorization) => {
        return (
          <React.Fragment key={authorization.code}>
            <Typography.Title level={5}>{authorization.description}</Typography.Title>
            {authorization.children &&
              authorization.children.map((child) => {
                return (
                  <FormSection name={child.code} key={`${authorization}.${child.code}`}>
                    <Checkbox
                      key={child.code}
                      value={child.code}
                      onChange={(e) => handleChange(e, child.code)}
                      checked={checked.includes(child.code)}
                    >
                      {checked.includes(child.code) ? (
                        <>
                          {child.description} <DownOutlined />
                        </>
                      ) : (
                        child.description
                      )}
                    </Checkbox>
                    {checked.includes(child.code) && renderNestedFields(child.code)}
                  </FormSection>
                );
              })}
            <br />
          </React.Fragment>
        );
      })}
    </>
  );
};

AuthorizationsInvolved.propTypes = propTypes;

const mapStateToProps = (state) => ({
  transformedProjectSummaryAuthorizationTypes: getTransformedProjectSummaryAuthorizationTypes(
    state
  ),
  dropdownProjectSummaryPermitTypes: getDropdownProjectSummaryPermitTypes(state),
  formattedProjectSummary: getFormattedProjectSummary(state),
  formValues: getFormValues(FORM.ADD_EDIT_PROJECT_SUMMARY)(state),
});

const mapDispatchToProps = (dispatch) => bindActionCreators({ change }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(AuthorizationsInvolved);
