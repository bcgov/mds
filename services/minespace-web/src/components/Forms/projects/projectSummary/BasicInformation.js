import React from "react";
import { Typography } from "antd";
import { Field } from "redux-form";
import { maxLength, required } from "@common/utils/Validate";
import { renderConfig } from "@/components/common/config";
import LinkedProjects from "@/components/Forms/projects/projectSummary/LinkedProjects";

const propTypes = {};

export const BasicInformation = () => {
  return (
    <>
      <Typography.Title level={3}>Basic Information</Typography.Title>
      <Field
        id="project_summary_title"
        name="project_summary_title"
        label="Project title"
        component={renderConfig.FIELD}
        validate={[maxLength(300), required]}
      />
      <Field
        id="proponent_project_id"
        name="proponent_project_id"
        label={
          <>
            Proponent project tracking ID (optional)
            <br />
            <span className="light--sm">
              If your company uses a tracking number to identify projects, please provide it here.
            </span>
          </>
        }
        component={renderConfig.FIELD}
        validate={[maxLength(20)]}
      />
      <Field
        id="project_summary_description"
        name="project_summary_description"
        label={
          <>
            Project overview <br />
            <span className="light--sm">
              {" "}
              Provide a 2-3 paragraph high-level description of your proposed project.
            </span>
          </>
        }
        component={renderConfig.AUTO_SIZE_FIELD}
        minRows={10}
        validate={[maxLength(4000), required]}
      />
      <LinkedProjects />
    </>
  );
};

BasicInformation.propTypes = propTypes;

export default BasicInformation;
