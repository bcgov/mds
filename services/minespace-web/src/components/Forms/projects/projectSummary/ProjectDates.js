import React from "react";
import PropTypes from "prop-types";
import { Field, formValueSelector } from "redux-form";
import { Typography } from "antd";
import { connect } from "react-redux";
import { dateNotBeforeOther, dateNotAfterOther } from "@common/utils/Validate";
import Callout from "@/components/common/Callout";
import * as FORM from "@/constants/forms";
import RenderDate from "@mds/common/components/forms/RenderDate";

const propTypes = {
  expected_permit_application_date: PropTypes.string,
  expected_draft_irt_submission_date: PropTypes.string,
  expected_permit_receipt_date: PropTypes.string,
};

const defaultProps = {
  expected_permit_application_date: undefined,
  expected_draft_irt_submission_date: undefined,
  expected_permit_receipt_date: undefined,
};

export const ProjectDates = (props) => {
  return (
    <>
      <Typography.Title level={3}>Project Dates</Typography.Title>
      <Callout
        message={
          <>
            These dates are for guidance and planning purposes only and do not reflect actual
            delivery dates. The{" "}
            <a
              target="_blank"
              rel="noopener noreferrer"
              alt="Major Mines Permitting Office"
              href="https://www2.gov.bc.ca/gov/content/industry/mineral-exploration-mining/permitting/major-mines-permitting-office"
            >
              Major Mines Office
            </a>{" "}
            will work with you on a more definitive schedule.
          </>
        }
      />
      <Field
        id="expected_draft_irt_submission_date"
        name="expected_draft_irt_submission_date"
        label="When do you anticipate submitting a draft Information Requirements Table?"
        placeholder="Please select date"
        component={RenderDate}
        validate={[dateNotAfterOther(props.expected_permit_application_date)]}
      />
      <Field
        id="expected_permit_application_date"
        name="expected_permit_application_date"
        label="When do you anticipate submitting a permit application?"
        placeholder="Please select date"
        component={RenderDate}
        validate={[dateNotBeforeOther(props.expected_draft_irt_submission_date)]}
      />
      <Field
        id="expected_permit_receipt_date"
        name="expected_permit_receipt_date"
        label="When do you hope to receive your permit/amendment(s)?"
        placeholder="Please select date"
        component={RenderDate}
        validate={[dateNotBeforeOther(props.expected_permit_application_date)]}
      />
      <Field
        id="expected_project_start_date"
        name="expected_project_start_date"
        label="When do you anticipate starting work on this project?"
        placeholder="Please select date"
        component={RenderDate}
        validate={[dateNotBeforeOther(props.expected_permit_receipt_date)]}
      />
    </>
  );
};

ProjectDates.propTypes = propTypes;
ProjectDates.defaultProps = defaultProps;

const selector = formValueSelector(FORM.ADD_EDIT_PROJECT_SUMMARY);
const mapStateToProps = (state) => ({
  expected_draft_irt_submission_date: selector(state, "expected_draft_irt_submission_date"),
  expected_permit_application_date: selector(state, "expected_permit_application_date"),
  expected_permit_receipt_date: selector(state, "expected_permit_receipt_date"),
});

export default connect(mapStateToProps)(ProjectDates);
