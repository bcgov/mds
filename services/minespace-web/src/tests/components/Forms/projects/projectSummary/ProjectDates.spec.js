import React from "react";
import { shallow } from "enzyme";
import { ProjectDates } from "@/components/Forms/projects/projectSummary/ProjectDates";

const dispatchProps = {};
const props = {};

const setupDispatchProps = () => {};

const setupProps = () => {
  props.expected_permit_application_date = undefined;
  props.expected_draft_irt_submission_date = undefined;
  props.expected_permit_receipt_date = undefined;
};

beforeEach(() => {
  setupDispatchProps();
  setupProps();
});

describe("ProjectDates", () => {
  it("renders properly", () => {
    const component = shallow(<ProjectDates {...dispatchProps} {...props} />);
    expect(component).toMatchSnapshot();
  });
});
