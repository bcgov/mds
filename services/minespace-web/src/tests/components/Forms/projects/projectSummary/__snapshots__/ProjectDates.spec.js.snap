// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDates renders properly 1`] = `
<Fragment>
  <ForwardRef
    level={3}
  >
    Project Dates
  </ForwardRef>
  <Callout
    message={
      <React.Fragment>
        These dates are for guidance and planning purposes only and do not reflect actual delivery dates. The
         
        <a
          alt="Major Mines Permitting Office"
          href="https://www2.gov.bc.ca/gov/content/industry/mineral-exploration-mining/permitting/major-mines-permitting-office"
          rel="noopener noreferrer"
          target="_blank"
        >
          Major Mines Office
        </a>
         
        will work with you on a more definitive schedule.
      </React.Fragment>
    }
    severity="info"
    style={Object {}}
  />
  <Field
    component={[Function]}
    id="expected_draft_irt_submission_date"
    label="When do you anticipate submitting a draft Information Requirements Table?"
    name="expected_draft_irt_submission_date"
    placeholder="Please select date"
    validate={
      Array [
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    id="expected_permit_application_date"
    label="When do you anticipate submitting a permit application?"
    name="expected_permit_application_date"
    placeholder="Please select date"
    validate={
      Array [
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    id="expected_permit_receipt_date"
    label="When do you hope to receive your permit/amendment(s)?"
    name="expected_permit_receipt_date"
    placeholder="Please select date"
    validate={
      Array [
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    id="expected_project_start_date"
    label="When do you anticipate starting work on this project?"
    name="expected_project_start_date"
    placeholder="Please select date"
    validate={
      Array [
        [Function],
      ]
    }
  />
</Fragment>
`;
